src/main/java/graphql/annotations/connection/EnhancedConnectionFetcher.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright 2016 Yurii Rashkovskii
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ */
+package graphql.annotations.connection;
+
+import graphql.relay.*;
+import graphql.schema.DataFetchingEnvironment;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Use this class in {@link GraphQLConnection} to do a real pagination,
+ * i.e you fetch each time the relevant data, you make the cursors and
+ * you decide if there are previous or next pages
+ *
+ * Note: If you are using the connection, the relevant dataFetcher must implements {@link PaginationDataFetcher}
+ * @param <T> the entity type that is paginated
+ */
+public class EnhancedConnectionFetcher<T> implements ConnectionFetcher<Connection<T>> {
+
+    private PaginationDataFetcher<T> paginationDataFetcher;
+
+    public EnhancedConnectionFetcher(PaginationDataFetcher<T> paginationDataFetcher) {
+        this.paginationDataFetcher = paginationDataFetcher;
+    }
+
+    @Override
+    public Connection<T> get(DataFetchingEnvironment environment) {
+        List<Edge<T>> edges = buildEdges(paginationDataFetcher.get(environment));
+        PageInfo pageInfo = getPageInfo(edges);
+        return new DefaultConnection<>(edges, pageInfo);
+    }
+
+    private PageInfo getPageInfo(List<Edge<T>> edges) {
+        ConnectionCursor firstCursor = edges.get(0).getCursor();
+        ConnectionCursor lastCursor = edges.get(edges.size()-1).getCursor();
+        return new DefaultPageInfo(
+                firstCursor,
+                lastCursor,
+                paginationDataFetcher.hasPreviousPage(firstCursor.getValue()),
+                paginationDataFetcher.hasNextPage(lastCursor.getValue())
+        );
```

Comment:
this seems a little weird in that you represent the "encoded cursors" back to the `paginationDataFetcher` that provided them.  Therefore to implement this one would need to "decode" the value it just encoded.  Not the end of the world but ...

Perhaps you need the `paginationDataFetcher` to pass back pagination info on the original fetch via a compound value intermediate value.

    class CompundData {
         List<T> data
         boolean hasNext
        boolean hasPrev
   }